File: modulo-01-fundamentos-arquitectura/codigo/src/main/java/pe/joedayz/microservicios/modulo01/app/OrderApplicationService.java
package pe.joedayz.microservicios.modulo01.app;

import pe.joedayz.microservicios.modulo01.ddd.order.Order;
import pe.joedayz.microservicios.modulo01.ddd.order.OrderRepository;
import pe.joedayz.microservicios.modulo01.ddd.shared.DomainEvent;
import pe.joedayz.microservicios.modulo01.patterns.outbox.OutboxStore;

/**
 * Servicio de aplicacion (caso de uso). Orquesta el agregado y la persistencia, pero
 * NO contiene reglas de negocio (esas viven en el agregado {@link Order}).
 *
 * <p>Aqui se ve el patron TRANSACTIONAL OUTBOX: en una unica "transaccion" guardamos el
 * pedido Y sus eventos de dominio en la outbox. Al ser atomico, es imposible tener un
 * pedido guardado sin su evento (evita el problema del dual-write).
 */
public class OrderApplicationService {

    private final OrderRepository orderRepository;
    private final OutboxStore outbox;

    public OrderApplicationService(OrderRepository orderRepository, OutboxStore outbox) {
        this.orderRepository = orderRepository;
        this.outbox = outbox;
    }

    /**
     * Persiste el estado del pedido y vuelca sus eventos de dominio a la outbox,
     * todo en la misma unidad de trabajo (simulada).
     */
    public void saveWithEvents(Order order) {
        // --- inicio transaccion (BEGIN) ---
        orderRepository.save(order);
        for (DomainEvent event : order.pullDomainEvents()) {
            outbox.add(event);
        }
        // --- fin transaccion (COMMIT) : pedido + eventos, atomico ---
    }
}

File: modulo-01-fundamentos-arquitectura/codigo-dotnet/src/Modulo01.Fundamentos/Application/OrderApplicationService.cs
using JoeDayz.Microservicios.Modulo01.Ddd.Orders;
using JoeDayz.Microservicios.Modulo01.Patterns.Outbox;

namespace JoeDayz.Microservicios.Modulo01.Application;

/// <summary>
/// Servicio de aplicacion (caso de uso). Orquesta agregado + persistencia, pero NO
/// contiene reglas de negocio (esas viven en <see cref="Order"/>).
/// Aqui se ve el TRANSACTIONAL OUTBOX: pedido y eventos se guardan en la MISMA
/// transaccion, de modo que es imposible tener un pedido sin su evento (dual-write).
/// </summary>
public sealed class OrderApplicationService(IOrderRepository orderRepository, OutboxStore outbox)
{
    public void SaveWithEvents(Order order)
    {
        // --- BEGIN TRANSACTION ---
        orderRepository.Save(order);
        foreach (var domainEvent in order.PullDomainEvents())
        {
            outbox.Add(domainEvent);
        }
        // --- COMMIT: pedido + eventos, atomico ---
    }
}
